File: 프로그래머스/3/84021. 퍼즐 조각 채우기/퍼즐 조각 채우기.py
from collections import deque

def is_movable(r, c, n):
    return 0 <= r < n and 0 <= c < n

def get_blocks(board, target):
    n = len(board)
    visited = [[False] * n for _ in range(n)]
    blocks = []
    
    dr = [-1, 1, 0, 0]
    dc = [0, 0, -1, 1]
    
    for r in range(n):
        for c in range(n):
            if board[r][c] == target and not visited[r][c]:
                block = []
                queue = deque([[r, c]])
                visited[r][c] = True
                
                while queue:
                    curr_r, curr_c = queue.popleft()
                    block.append([curr_r, curr_c])
                    
                    for i in range(4):
                        nr, nc = curr_r + dr[i], curr_c + dc[i]
                        
                        if is_movable(nr, nc, n) and not visited[nr][nc] and board[nr][nc] == target:
                            visited[nr][nc] = True
                            queue.append([nr, nc])
                
                blocks.append(block)
    return blocks

def normalize(block):
    # 최소 r과 최소 c 찾기
    min_r = min(r for r, c in block)
    min_c = min(c for r, c in block)
    
    # (0, 0) 기준으로 이동 후 정렬
    normalized = sorted([[r - min_r, c - min_c] for r, c in block])
    return normalized

def rotate(block):
    # (r, c) -> (c, -r) 90도 회전
    rotated = [[c, -r] for r, c in block]
    return normalize(rotated)

def solution(game_board, table): 
    blanks = [normalize(b) for b in get_blocks(game_board, 0)]
    puzzles = [normalize(p) for p in get_blocks(table, 1)]
    
    puzzle_used = [False] * len(puzzles)
    
    answer = 0
    
    for blank in blanks:
        found = False
        for i in range(len(puzzles)):
            if puzzle_used[i]:
                continue
            if len(blank) != len(puzzles[i]):
                continue
                
            current_puzzle = puzzles[i]
            
            # 0, 90, 180, 270도 회전하며 대조
            for _ in range(4):
                if blank == current_puzzle: # 파이썬은 리스트끼리 == 비교 가능!
                    answer += len(blank)
                    puzzle_used[i] = True
                    found = True
                    break
                current_puzzle = rotate(current_puzzle)
            
            if found:
                break
                
    return answer

File: 프로그래머스/3/84021. 퍼즐 조각 채우기/퍼즐 조각 채우기.js

const dr = [-1, 1, 0, 0];
const dc = [0, 0, -1, 1];

function isMovable(r, c, n) {
    return r >= 0 && r < n && c >= 0 && c < n;
}

// target: 0 또는 1
function getBlocks(board, target) {
    const n = board.length;
    const visited = Array.from({ length: n }, () => Array(n).fill(false));
    const blocks = [];
    
    for (let r = 0; r < n; r++) {
        for (let c = 0; c < n; c++) {
            if (board[r][c] === target && !visited[r][c]) {
                const block = [];
                const queue = [[r, c]];
                visited[r][c] = true;
                
                let head = 0;
                while (head < queue.length) {
                    const [curR, curC] = queue[head++];
                    block.push([curR, curC]);

                    for (let i = 0; i < 4; i++) {
                        const nr = curR + dr[i];
                        const nc = curC + dc[i];

                        if (isMovable(nr, nc, n) && !visited[nr][nc] && board[nr][nc] === target) {
                            visited[nr][nc] = true;
                            queue.push([nr, nc]);
                        }
                    }
                }
                blocks.push(block);
            }
        }
    }
    return blocks;
}

// 조각들을 표준화하기
function normalize(block) {
    let minR = Infinity;
    let minC = Infinity;
    
    for (const [r, c] of block) {
        minR = Math.min(minR, r);
        minC = Math.min(minC, c);
    }

    // 정렬 기준: 행(r)이 작은 순서 -> 행이 같다면 열(c)이 작은 순서
    return block
        .map(([r, c]) => [r - minR, c - minC])
        .sort((a, b) => a[0] !== b[0] ? a[0] - b[0] : a[1] - b[1]);
}

// 조각들을 회전시키는 함수
function rotate(block) {
    const rotatedBlock = block.map(([r, c]) => [c, -r]);
    return normalize(rotatedBlock);
}
// 조각이 딱 맞는지 체크하는 함수
function isSame(block1, block2) {
    if (block1.length !== block2.length) return false;
    for (let i = 0; i < block1.length; i++) {
        if (block1[i][0] !== block2[i][0] || block1[i][1] !== block2[i][1]) {
            return false;
        }
    }
    return true;
}

function solution(game_board, table) {
    const blanks = getBlocks(game_board, 0).map(b => normalize(b));
    const puzzles = getBlocks(table, 1).map(p => normalize(p));
    
    const puzzleUsed = Array(puzzles.length).fill(false);
    
    let answer = 0;
    
    for (const blank of blanks) {
        let found = false;
        
        for (let i = 0; i < puzzles.length; i++) {
            if (puzzleUsed[i]) continue;
            if (blank.length !== puzzles[i].length) continue;
        
            let currentPuzzle = puzzles[i];
        
            // 조각을 0도, 90도, 180도, 270도 회전시키며 대조
            for (let r = 0; r < 4; r++) {
                if (isSame(blank, currentPuzzle)) {
                    answer += blank.length;
                    puzzleUsed[i] = true;
                    found = true;
                    break;
                }
                currentPuzzle = rotate(currentPuzzle); // 90도 회전
            }
            
            if (found) break;
        }
    }
    
    return answer;
}